Measure unsubscribed by type [MAILPOET-4733]

samnajian opened the pull request:
## Description

This PR adds different methods to unsubscribe records based on method and exposes these stats to the reporter.

## Code review notes

_N/A_

## QA notes

Verify the following are exposed:
```
'Unsubscribe > Total in last 6 months'
'Unsubscribe > Count in last 6 months with method: Link'
'Unsubscribe > Count in last 6 months with method: 1 Click'
```

## Linked tickets

[MAILPOET-4733]


[MAILPOET-4733]: https://mailpoet.atlassian.net/browse/MAILPOET-4733?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

samnajian committed:
Add method of tracking when tracking unsubscriptions

[MAILPOET-4733]

samnajian committed:
Augment track persister test with method as meta

[MAILPOET-4733]

samnajian committed:
Add test to assert correct unsubscription method is saved

[MAILPOET-4733]

samnajian committed:
Expose unsubscribe analytics properperties

[MAILPOET-4733]

samnajian committed:
Add integration test for unsubscibe properties

[MAILPOET-4733]

costasovo commented on mailpoet/lib/Analytics/UnsubscribeReporter.php:
```diff
@@ -0,0 +1,48 @@
+<?php
```
Please add `declare(strict_types = 1);` we want to add all new PHP files with the strict types enabled.

costasovo commented on mailpoet/tests/integration/Analytics/UnsubscribeAnalyticsTest.php:
```diff
@@ -0,0 +1,69 @@
+<?php
```
Please add `declare(strict_types = 1);` we want to add all new PHP files with the strict types enabled.

costasovo commented on mailpoet/lib/Analytics/UnsubscribeReporter.php:
```diff
@@ -0,0 +1,48 @@
+  /**
+   * @param string|null $methodKey
+   */
+  private function getMethodName($methodKey): string {
```
I think you can use a typehint `?string` here.

costasovo commented on mailpoet/lib/Analytics/UnsubscribeReporter.php:
```diff
@@ -0,0 +1,48 @@
+      return 'Link';
+    }
+
+    return 'Unknown';
```
Just an idea. We already [track admin names when tracking unsubscribers](https://github.com/mailpoet/mailpoet/blob/76048224d112a46c1a1d60873694798e351600f5/mailpoet/lib/Subscribers/SubscriberSaveController.php#L139) so maybe we could distingush some value in meta as method `admin` and `null` as `Unknown`.

costasovo reviewed (changes requested):
Looks good! I found a couple of tiny details. See my comments.

I have one question, though. I see we use `meta` column for saving the unsubscribe type as a plain string. In two other tables where we have `meta` (scheduled_tasks, sending_queues) we save serialized array, which allows us to store whatever data and extend the data with new properties.
So I was about to ask you to use a different column, but then I noticed the type is just varchar, and [we already save plain string (user name) there](https://github.com/mailpoet/mailpoet/blob/76048224d112a46c1a1d60873694798e351600f5/mailpoet/lib/Subscribers/SubscriberSaveController.php#L139). So, in this case, it seems we just use the meta column differently.
I just want to make sure you are aware that we already save the names of users who set a subscriber as unsubscribed into the column. Did you consider using a new column for storing the type of unsubscribe?

samnajian commented:
> Looks good! I found a couple of tiny details. See my comments.
> 
> I have one question, though. I see we use `meta` column for saving the unsubscribe type as a plain string. In two other tables where we have `meta` (scheduled_tasks, sending_queues) we save serialized array, which allows us to store whatever data and extend the data with new properties. So I was about to ask you to use a different column, but then I noticed the type is just varchar, and [we already save plain string (user name) there](https://github.com/mailpoet/mailpoet/blob/76048224d112a46c1a1d60873694798e351600f5/mailpoet/lib/Subscribers/SubscriberSaveController.php#L139). So, in this case, it seems we just use the meta column differently. I just want to make sure you are aware that we already save the names of users who set a subscriber as unsubscribed into the column. Did you consider using a new column for storing the type of unsubscribe?

Actually, at first I thought I'd use a new column for unsubscribe type, but then realized we are not using `meta` column much, so it felt like introducing a new column would be too much, so left this decision to the PR, I'm still not convinced introducing a new column is a good idea, I'd get this discussion sync with you.

samnajian committed:
Introduce method column to statistics_unsubscribes table

[MAILPOET-4733]

samnajian committed:
Adjust strict_types & method typehint based on PR feedback

[MAILPOET-4733]

samnajian committed:
Set missing unsubscribe method for failing tests

[MAILPOET-4733]

costasovo commented on mailpoet/lib/Entities/StatisticsUnsubscribeEntity.php:
```diff
@@ -108,4 +114,11 @@ public function setMeta(string $meta) {
+  /**
+   * @return string
+   */
+  public function getMethod() {
```
Could you please add return type hint `public function getMethod(): string`?
Actually, with the return type hint, we don't need PHP doc for `@return` but I see it is there for all other getters. Probably added by ide.

costasovo commented on mailpoet/lib/Analytics/UnsubscribeReporter.php:
```diff
@@ -31,9 +31,6 @@ public function getProperties(): array {
-  /**
-   * @param string|null $methodKey
-   */
   private function getMethodName($methodKey): string {
```
I see you removed PHP doc but you didn't add the typehint.
I think it should be `private function getMethodName(?string $methodKey): string {`

costasovo reviewed (changes requested):
Thank you for adding the changes!
I feel a bit like a nitpick, but could you please make the two tiny type-hint improvements I suggest in the new comments?

costasovo reviewed (dismissed):
Thank you!

Aschepikov reviewed (commented):
Hi, I have an issue with subscribers edit info.
When I am trying to press on an edit of any subscriber it returns 400 error
`error
: 
"bad_request"
message
: 
"An exception occurred while executing 'SELECT t0.source AS source_1, t0.meta AS meta_2, t0.method AS method_3, t0.id AS id_4, t0.created_at AS created_at_5, t0.newsletter_id AS newsletter_id_6, t0.queue_id AS queue_id_7, t0.subscriber_id AS subscriber_id_8 FROM wp_mailpoet_statistics_unsubscribes t0 WHERE t0.subscriber_id = ? ORDER BY t0.created_at DESC' with params [10896]:\n\nSQLSTATE[42S22]: Column not found: 1054 Unknown column 't0.method' in 'field list'"`
`<html>
<body>
<!--StartFragment--><div aria-expanded="true" tabindex="-1" style="box-sizing: border-box; min-width: 0px; min-height: 0px; flex: 0 0 auto; color: rgb(32, 33, 36); font-family: consolas, &quot;lucida console&quot;, &quot;courier new&quot;, monospace; font-size: 12px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(236, 241, 248); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span class="source-code" style="box-sizing: border-box; min-width: 0px; min-height: 0px; font-family: var(--source-code-font-family); white-space: pre-wrap; font-size: var(--source-code-font-size) !important; line-height: 1.2;"><span class="console-message-anchor" style="box-sizing: border-box; min-width: 0px; min-height: 0px; float: right; text-align: right; max-width: 100%; margin-left: 4px;"><span class="devtools-link" role="link" tabindex="-1" title="https://gs.qawp.net/wp-admin/load-scripts.php?c=1&amp;load%5Bchunk_0%5D=jquery-core,jquery-migrate,utils&amp;ver=6.1.1:2" style="box-sizing: border-box; min-width: 0px; min-height: 0px; color: var(--color-text-secondary); text-decoration: underline; cursor: pointer; word-break: break-all;"><br class="Apple-interchange-newline">load-scripts.php?c=1<span class="devtools-link-ellipsis" style="box-sizing: border-box; min-width: 0px; min-height: 0px;">…</span>e,utils&amp;ver=6.1.1:2</span> <devtools-request-link-icon class="resource-links" style="display: inline-block; white-space: nowrap; color: inherit; font-size: inherit; font-family: inherit; box-sizing: border-box; min-width: 0px; min-height: 0px; vertical-align: bottom;"><span tabindex="0" class=" link " style="cursor: pointer;"><devtools-icon title="Click to open the network panel and show request for URL: https://gs.qawp.net/wp-admin/admin-ajax.php" style="display: inline-block; white-space: nowrap; vertical-align: middle;"><div class="icon-basic" style="width: 16px; height: 16px; display: block; -webkit-mask-image: url(&quot;devtools://devtools/bundled/devtools-frontend/front_end/Images/network_panel_icon.svg&quot;); -webkit-mask-position: center center; -webkit-mask-repeat: no-repeat; -webkit-mask-size: 99%; background-color: var(--icon-color, var(--color-link));"></div></devtools-icon></span></devtools-request-link-icon> </span><span class="console-message-text" style="box-sizing: border-box; min-width: 0px; min-height: 0px; color: var(--override-error-text-color) !important;">POST <span class="devtools-link" role="link" tabindex="-1" style="box-sizing: border-box; min-width: 0px; min-height: 0px; color: var(--color-text-secondary); text-decoration: underline; cursor: pointer; word-break: break-all;">https://gs.qawp.net/wp-admin/admin-ajax.php</span> 400</span></span></div><div class="" role="group" style="box-sizing: border-box; min-width: 0px; min-height: 0px; flex: 0 0 auto; color: rgb(32, 33, 36); font-family: consolas, &quot;lucida console&quot;, &quot;courier new&quot;, monospace; font-size: 12px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(236, 241, 248); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><span class="monospace" style="display: inline-block; --monospace-font-size:12px; --monospace-font-family:consolas,lucida console,courier new,monospace; --source-code-font-size:12px; --source-code-font-family:consolas,lucida console,courier new,monospace; box-sizing: border-box; min-width: 0px; min-height: 0px; font-family: var(--monospace-font-family); font-size: var(--monospace-font-size) !important;">

  | send | @ | load-scripts.php?c=1…e,utils&ver=6.1.1:2
-- | -- | -- | --
  | ajax | @ | load-scripts.php?c=1…e,utils&ver=6.1.1:2
  | s.ajax.s.ajax | @ | load-scripts.php?c=1…e,utils&ver=6.1.1:5
  | S.<computed> | @ | load-scripts.php?c=1…e,utils&ver=6.1.1:2
  | request | @ | commons.js?ver=4.0.0:9
  | post | @ | commons.js?ver=4.0.0:9
  | (anonymous) | @ | admin.js?ver=4.0.0:2
  | componentDidMount | @ | admin.js?ver=4.0.0:2
  | As | @ | commons.js?ver=4.0.0:9
  | kl | @ | commons.js?ver=4.0.0:9
  | t.unstable_runWithPriority | @ | commons.js?ver=4.0.0:9
  | ji | @ | commons.js?ver=4.0.0:9
  | Cl | @ | commons.js?ver=4.0.0:9
  | fl | @ | commons.js?ver=4.0.0:9
  | (anonymous) | @ | commons.js?ver=4.0.0:9
  | t.unstable_runWithPriority | @ | commons.js?ver=4.0.0:9
  | ji | @ | commons.js?ver=4.0.0:9
  | zi | @ | commons.js?ver=4.0.0:9
  | Vi | @ | commons.js?ver=4.0.0:9
  | Fe | @ | commons.js?ver=4.0.0:9
  | $t | @ | commons.js?ver=4.0.0:9

</span></div><!--EndFragment-->
</body>
</html>`

When I a, pressing edit
![image](https://user-images.githubusercontent.com/113185437/203776130-5c103052-d6f1-4e57-9b0e-b38b538bfae7.png)
And then
![image](https://user-images.githubusercontent.com/113185437/203776239-2542d447-5f39-4121-8f26-e4312b1a8115.png)
![image](https://user-images.githubusercontent.com/113185437/203776383-f2e48b2f-82be-451c-b94c-3e355e6e2996.png)

Aschepikov reviewed (commented):
Sorry, I think this issue was connected to my site

JanJakes committed:
Use helper methods for method column migration

This fixes issues with MailPoet in multiple schemas on the same MySQL instance as well
as simplifies the code. The migration was also renamed to a newer date. Additionally, the
edit of the initial MailPoet migration was reverted as this is not needed.

[MAILPOET-4733]

JanJakes commented:
@costasovo noticed me that with the migration here we'd run into similar issues as fixed in https://github.com/mailpoet/mailpoet/pull/4559. Since @samnajian is AFK, after talking to @Aschepikov, I rebased this on `trunk`, and fixed the migration to reuse the same helpers we created in the previous fix.

costasovo reviewed (approved):
LGTM!

Aschepikov reviewed (commented):
Hi, I tested the footer link method and it is working correctly 
![image](https://user-images.githubusercontent.com/113185437/203946950-d1c61f87-d5f0-4227-96a7-6c5f3adb401b.png)
But when I was testing 1 click unsubscribe it wasn't working
For example, I unsubscribed here for this user
![image](https://user-images.githubusercontent.com/113185437/203947161-f71723de-2b3b-4f76-b7ae-a4afed074647.png)
But he is still subscribed
![image](https://user-images.githubusercontent.com/113185437/203947315-babe4f03-d207-4e68-a775-d00fb0ab6a9a.png)

On the prod if we are unsubscribing here
![image](https://user-images.githubusercontent.com/113185437/203947409-b1d8af31-c454-4249-a776-a7232de46f30.png)
We are unsubscribed here too
![image](https://user-images.githubusercontent.com/113185437/203947460-4a606a63-a5af-4569-b301-dfadbdf1a01f.png)

samnajian commented:
@Aschepikov just checked with the latest build artifact of this PR and worked correctly, let's check together or explain how I should try to reproduce it.

Aschepikov reviewed (approved):
Issue connected with the site again. Verified, good job!
